src/lib/portal-api.ts

```diff
@@ -0,0 +1,150 @@
+          const thisUsersOfferings = res.body as IPortalOffering[];
+          const clueOfferings = thisUsersOfferings.filter(offering => isClueAssignment(offering));
+          resolve(clueOfferings);
+        }
+      });
+    }
+    else {
+      resolve([]);
+    }
+  });
+};
+
+export const getProblemIdForAuthenticatedUser = (rawPortalJWT: string, urlParams?: QueryParams) => {
+  return new Promise<string|undefined>((resolve, reject) => {
+    if (urlParams && urlParams.offering) {
+      superagent
+      .get(urlParams.offering)
+      .set("Authorization", `Bearer/JWT ${rawPortalJWT}`)
+      .end((err, res) => {
+        if (err) {
+          reject(getErrorMessage(err, res));
+        } else {
+          const activityUrl = ((res.body || {}).activity_url) || "";
+          const [ignore, query, ...rest] = activityUrl.split("?");
+          const params = queryString.parse(query);
+          resolve(params.problem as string);
+        }
+      });
+    }
+    else {
+      resolve(undefined);
+    }
+  });
+};
+
+// The portals native format of API returns
+interface IPortalOffering {
+  id: number;
+  clazz: string;
+  clazz_id: number;
+  activity: string;
+  activity_url: string;
+  external_report?: IPortalReport | null;
+  external_reports?: IPortalReport[];
+}
+
+// Extracts the problem ordinal from the activity_url. An activity_url is part
+// of what the portal returns as an offering and has the problem ordinal at the
+// end, e.g. "https://collaborative-learning.concord.org/branch/master/index.html?problem=3.1"
+function getProblemOrdinal(offering: IPortalOffering) {
+  const defaultOrdinal = "x.x.x";
+  const problemRegEx = /problem=(\d\.\d)/;
+  const result = problemRegEx.exec(offering.activity_url);
+  if (result && result[1]) {
+    return (result[1] as string) || defaultOrdinal;
+  }
+  return defaultOrdinal;
+}
+
+export function getClueClassOfferings(portalOfferings: IPortalOffering[], urlParams?: QueryParams) {
+  const result = [] as IClueClassOffering[];
+  const addOffering = (offering: IPortalOffering) => {
+    if (isClueAssignment(offering) && urlParams) {
+      let newLocationUrl = "";
+      if (urlParams && urlParams.class && urlParams.offering && urlParams.reportType && urlParams.token) {
+        newLocationUrl =
+          `?class=${urlParams.class.replace(/\/classes\/.*$/, `/classes/${offering.clazz_id}`)}` +
+          `&offering=${urlParams.offering.replace(/\/offerings\/.*$/, `/offerings/${offering.id}`)}` +
+          `&reportType=${urlParams.reportType}` +
+          `&token=${urlParams.token}`;
+      }
+      result.push({
+        className: offering.clazz,
+        problemOrdinal: getProblemOrdinal(offering),
+        offeringId: `${offering.id}`,
+        location: newLocationUrl
+      });
+    }
+  };
+  portalOfferings.forEach(addOffering);
+  return sortOfferings(result);
+}
+
+// Sorts the offerings by class name (alphabetically) and then by problem ordinal
+// within each class.
```

Comment:
Should unit be part of this sort?

Reply:
Sorting by unit is an interesting idea, but right now we are filtering on same unit problems.